File: Binary_Search/347.Top-K-Frequent-Elements/347.Top-K-Frequent-Elements.ts
function topKFrequent(nums: number[], k: number): number[] {
  const ret: number[] = [];
  const notebook = new Map();

  for (let i = 0; i < nums.length; i++) {
    notebook.set(nums[i], (notebook.get(nums[i]) || 0) + 1);
  }

  // 找K，具体的频率是多少
  let left = 0;
  let right = nums.length;
  while (left < right) {
    const f = (left + right) >> 1;
    if (countFreqGreaterOrEqual(f) >= k) {
      left = f;
    } else {
      right = f - 1;
    }
  }
  const f = left;

  for (const [k, v] of notebook) {
    if (v >= f) {
      ret.push(k);
    }
  }
  return ret;
  function countFreqGreaterOrEqual(f: number) {
    let counter = 0;
    for (const [, v] of notebook) {
      if (v >= f) {
        counter++;
      }
    }
    return counter;
  }
}

File: Binary_Search/347.Top-K-Frequent-Elements/347.Top-K-Frequent-Elements.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        notebook = dict(Counter(nums))
        # 找到第 K 个的元素的频率到底是多少
        left = 0
        right = len(nums)
        while left < right:
            middle = math.ceil((left + right) / 2)
            if self.freqGreaterOrEqualCount(middle, notebook) >= k:
                left = middle
            else:
                right = middle - 1

        result = []
        for num in notebook:
            if notebook[num] >= right:
                result.append(num)
        return result
                
    def freqGreaterOrEqualCount(self, times, notebook):
        count = 0
        for num in notebook:
            if notebook[num] >= times:
                count += 1
        return count
